nixos: turn vmWithBootLoader into option (`nixos-rebuild build-vm`)

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

 - Clean up by not repeating `evalModules` calls or `vm` definitions.
 - Allow users to affect the `nixos-rebuild build-vm` configuration directly, by setting configs in the new options.
 - Make the `vm` and `vmWithBootloader` attrs less special. These are now in regular config values.
 - Fix the type of `system.build`.

This is similar to https://github.com/NixOS/nixpkgs/pull/144094

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth committed:
nixos/default.nix: Use extendModules

roberth committed:
nixos: Move build-vm into virtualisation.vmVariant

... which is like a specialisation, but for nixos-rebuild build-vm

ncfavier commented:
I think `nixos-rebuild` would have to be updated to handle `build-vm*` with flakes correctly.

ncfavier commented on nixos/default.nix:
```diff
@@ -9,35 +9,14 @@ let
+  vm = eval.config.virtualisation.vmVariant.system.build.vm;
 
-  vmWithBootLoader = vmWithBootLoaderConfig.system.build.vm;
+  vmWithBootLoader = eval.config.virtualisation.vmVariantWithBootLoader.system.build.vm;
```
Maybe we should deprecate those and use the `vmVariant`s directly in `nixos-rebuild`?

roberth replied:
Maybe at some point after a long transitioin period. They're canon: `nix-build <nixpkgs/nixos> -A vm` is a thing people do.

ncfavier replied:
Fair, that's a shorthand worth having.

ncfavier commented on nixos/lib/eval-config.nix:
```diff
@@ -92,7 +92,7 @@ in withWarnings {
   # Merge the option definitions in all modules, forming the full
   # system configuration.
-  inherit (nixosWithUserModules) config options _module type;
+  inherit (nixosWithUserModules) config options _module type extendModules;
```
Since this is the whole attrset, maybe just `nixosWithUserModules // { ... }` ?

roberth committed:
flake.nix: Deduplicate vmConfig, vmWithBootloaderConfig

ncfavier commented on flake.nix:
```diff
@@ -59,16 +41,16 @@
-                    vm = vmConfig.system.build.vm;
-                    vmWithBootLoader = vmWithBootLoaderConfig.system.build.vm;
+                    vm = config.virtualisation.vmVariant.system.build.vm;
+                    vmWithBootLoader = config.virtualisation.vmVariantWithBootLoader.system.build.vm;
```
Doesn't this cause infinite recursion?

roberth replied:
No, `extendModules` produces a new config.

I am concerned about this attr colliding though. What if someone invokes `nixosSystem` for the purpose of creating an image with the `qemu-vm.nix` module that they imported themselves.

ncfavier replied:
I've confirmed that it works, but I don't get why.

`extendModules` reuses the modules that were passed to `evalModules`, in this case one containing `system.build.vm = config.virtualisation.vmVariant.system.build.vm;`, so why doesn't `qemu-vm.nix` collide with that?

roberth replied:
`qemu-vm.nix` isn't in `module-list.nix`, so it's not part of the normal config unless you `imports` it manually.

It lines up almost too perfectly.

ncfavier replied:
But it is imported [here](https://github.com/NixOS/nixpkgs/pull/151082/files#diff-50176a86aa4300bb428500e2ff1d920bb67b8a216cf3e173e2896b5801645eaaR8-R10), I'd expect things to clash there

roberth replied:
It's because of legacy `types.attrs` :fearful:

ncfavier replied:
Ah... guess the `mkDefault` solution is fine

roberth committed:
nixos/lib/eval-config.nix: Return all of evalModules return attrs

We were exposing everything pointwise anyway.
If any new attrs are added, there's a good chance we'll want to
expose them anyway.

roberth committed:
flake.nix: lib.nixosSystem: Set system.build.vm* with lib.mkDefault

This will help anyone who imports the qemu module themselves, to
avoid a collision.

roberth committed:
nixos: Make system.build a lazyAttrsOf unspecified

Legacy types.attrs has really bad merging behavior and does not
support priorities.

f build

roberth committed:
nixos: Make system.build.vm a standard attribute based on vmVariant

roberth committed:
nixos: Add release note about vmVariant

ncfavier commented on flake.nix:
```diff
@@ -59,16 +41,11 @@
               in
               map addModuleDeclarationFile modules ++ [
-                {
+                ({ config, ... }: {
```
Not needed anymore

roberth committed:
flake.nix: Remove redundant module lambda

ncfavier commented on nixos/default.nix:
```diff
@@ -9,35 +9,14 @@ let
+  vm = eval.config.virtualisation.vmVariant.system.build.vm;
 
-  vmWithBootLoader = vmWithBootLoaderConfig.system.build.vm;
+  vmWithBootLoader = eval.config.virtualisation.vmVariantWithBootLoader.system.build.vm;
```
```nix
  inherit (eval.config.system.build) vm vmWithBootLoader;
```

ncfavier reviewed (approved):
Looks great now!

ncfavier commented:
Maybe we can add a `boot.isVM` option, in the spirit of `boot.isContainer`?

roberth commented:
> Maybe we can add a `boot.isVM` option, in the spirit of `boot.isContainer`?

Not a fan of such options, particularly when they're intended to remove things after the fact. The current solution is already more in the spirit of https://github.com/NixOS/nixpkgs/pull/148456 so I would consider depending on it to be somewhat of a regression. That said, I do think a use case may arise, but I'm not eager to add this option until we know for sure that we need it.

roberth commented:
> Looks great now!

Thanks for the fast paced collaboration / review!

ncfavier commented:
> Not a fan of such options, particularly when they're intended to remove things after the fact. The current solution is already more in the spirit of #148456 so I would consider depending on it to be somewhat of a regression. That said, I do think a use case may arise, but I'm not eager to add this option until we know for sure that we need it.

OK!

> > Looks great now!
> 
> Thanks for the fast paced collaboration / review!

NP!

dasJ commented:
Is this ready to get merged?

ncfavier commented:
I think so.

ncfavier commented:
I've tested that `nixos-rebuild build-vm` and `build-vm-with-bootloader` work fine both with flakes and non-flakes. This is OK to merge.

ncfavier commented:
This conflicts with https://github.com/NixOS/nixpkgs/pull/149532 in some way (tested before/after) and causes the manual to [fail to evaluate](https://github.com/NixOS/nixpkgs/runs/4820425506?check_suite_focus=true#step:5:161).

roberth commented:
Created https://github.com/NixOS/nixpkgs/pull/155022 but can't merge in the coming ~2h :( g2g

roberth commented:
Perhaps the attributes could be renamed, so that they don't trigger such logic.

> I'm not sure what other mechanisms there were

None that I know of and none documented here in the manual at least https://nixos.org/manual/nixos/stable/#sec-changing-config

ncfavier commented:
I am in favour of renaming the attributes, perhaps to something like `virtualisation.build.vm*` ?

Atemu commented:
This is also very non-trivial/near impossible to work around as `if then else` doesn't play well with the module system and an ugly `mkMerge` of `mkIf`s still fails because `virtualisation.vmVariant` (guarded behind a false `mkIf` statement!) doesn't exist in older Nixpkgs and the module system still complains.   

I finally found a workaround (`_module.check = false;` and `virtualisation.vmVariant = lib.mkIf ...`) but that took quite a while to debug.

ElvishJerricco commented:
Has this broken the virtualisation options in qemu-vm.nix? I can't set e.g. `virtualisation.cores = 4;` and use `nix-build ./nixpkgs/nixos --arg configuration ./configuration.nix -A vm`. I get ``error: The option `virtualisation.cores' does not exist.``

jackinloadup commented:
I'm running into the same issue as @ElvishJerricco but either I don't understand the proposed solution or it only works within nixpkgs. A related issue https://github.com/NixOS/nixpkgs/issues/59219 proposes a similar solution but is impure with `<nixpkgs>`. Is this import required by design? It seems like this file should already be imported at  https://github.com/NixOS/nixpkgs/blob/55c742cabea5b6290da8b9ffbc5ac26cf4741f76/nixos/modules/virtualisation/build-vm.nix#L8-L10 as part of  `nixos-rebuild build-vm`.

I could be misunderstanding as I'm still a nix beginner. My goal is to test my nixosConfigurations locally while tweaking them so I don't need to be near the target machine to see changes to the boot process.

roberth commented:
@jackinloadup I think ElvishJerricco's problem was unrelated to this PR.

`nixos-rebuild build-vm` should just work. If your configuration isn't quite compatible with the default vm variant of your configuration, you may have to add some overrides in the `virtualisation.vmVariant` option. It takes any NixOS option; for example: `virtualisation.vmVariant.services.nginx.enable = lib.mkForce false;`

If you know of a realistic option that needs to be overridden for `nixos-rebuild build-vm`, it'd be great to add it to the manual.
